File: src/main/java/com/rekeningrijden/simulation/Simulation/MessageProducer.java
package com.rekeningrijden.simulation.Simulation;

import com.rekeningrijden.europe.dtos.TransLocationDto;
import org.apache.log4j.Logger;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;

public class MessageProducer {
    private final static Logger logger = Logger.getLogger(MessageProducer.class);

    private Gateway SimulationToItaly;
    private Gateway SimulationToGermany;
    private Gateway SimulationToTheNetherlands;
    private Gateway SimulationToBelgium;
    private Gateway SimulationToFinland;

    public MessageProducer(){
        startSimulationToBelgium();
        startSimulationToFinland();
        startSimulationToGermany();
        startSimulationToItaly();
        startSimulationToTheNetherlands();
    }

    public void sendTransLocation(String countryCode, TransLocationDto dto) throws Exception {
        JSONObject jsonObj = new JSONObject(dto);

        switch (countryCode){
            case "IT":
                SimulationToItaly.channel.basicPublish("", "SimulationToItaly", null, convertPayLoadToBytes(jsonObj.toString()));
                break;
            case "DE":
                SimulationToItaly.channel.basicPublish("", "SimulationToGermany", null, convertPayLoadToBytes(jsonObj.toString()));
                break;
            case "NL":
                SimulationToItaly.channel.basicPublish("", "SimulationToTheNetherlands", null, convertPayLoadToBytes(jsonObj.toString()));
                break;
            case "BE":
                SimulationToItaly.channel.basicPublish("", "SimulationToBelgium", null, convertPayLoadToBytes(jsonObj.toString()));
                break;
            case "FI":
                SimulationToItaly.channel.basicPublish("", "SimulationToFinland", null, convertPayLoadToBytes(jsonObj.toString()));
                break;
            default:
                throw new Exception();
        }
        logger.debug("(" + countryCode + ") Payload has been sent to the queue");
    }

    private byte[] convertPayLoadToBytes(String payload) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ObjectOutputStream writter = new ObjectOutputStream(baos);
        writter.writeObject(payload);
        return baos.toByteArray();
    }

    private void startSimulationToItaly(){
        try {
            SimulationToItaly = new Gateway();
            SimulationToItaly.connect();
            SimulationToItaly.channel.queueDeclare("SimulationToItaly", false, false, false, null);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void startSimulationToGermany(){
        try {
            SimulationToGermany = new Gateway();
            SimulationToGermany.connect();
            SimulationToGermany.channel.queueDeclare("SimulationToGermany", false, false, false, null);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void startSimulationToTheNetherlands(){
        try {
            SimulationToTheNetherlands = new Gateway();
            SimulationToTheNetherlands.connect();
            SimulationToTheNetherlands.channel.queueDeclare("SimulationToTheNetherlands", false, false, false, null);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void startSimulationToBelgium(){
        try {
            SimulationToBelgium = new Gateway();
            SimulationToBelgium.connect();
            SimulationToBelgium.channel.queueDeclare("SimulationToBelgium", false, false, false, null);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void startSimulationToFinland(){
        try {
            SimulationToFinland = new Gateway();
            SimulationToFinland.connect();
            SimulationToFinland.channel.queueDeclare("SimulationToFinland", false, false, false, null);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/rekeningrijden/simulation/simulation/MessageProducer.kt
package com.rekeningrijden.simulation.simulation

import com.fasterxml.jackson.databind.ObjectMapper
import com.rabbitmq.client.AMQP
import com.rekeningrijden.europe.dtos.TransLocationDto
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.io.ByteArrayOutputStream
import java.io.ObjectOutputStream

@Service
class MessageProducer {
    private val mapper = ObjectMapper()
    private val gateways = mutableMapOf<String, Gateway>()

    init {
        val countries = arrayOf("BE", "DE", "FI", "IT", "NL")

        countries.forEach {
            val host = System.getProperty(createHostProperty(it))

            if (host == null)
                logger.error("Could not connect with the $it queue. If this is intended, please ignore this message. Otherwise, add the queue via a command line property:\n    -D${createQueueName(it)}=<ip-address>")
            else
                gateways[it] = createGateway(it, host)
        }

        if (gateways.count() == 0) {
            System.exit(1)
            logger.error("There are no RabbitMQ instances to be connected to. Please supply at least one valid URL.")
        }
    }

    fun sendTransLocation(countryCode: String, dto: TransLocationDto) {
        val payload = mapper.writeValueAsString(dto)
        val bytePayload = convertPayLoadToBytes(payload)

        /**
         * The message will be marked with delivery mode 2, this means that the
         * message will be persisted on disk by RabbitMQ so that it will not be
         * lost.
         */
        val props = AMQP.BasicProperties
            .Builder()
            .contentType("application/x-java-serialized-object")
            .deliveryMode(2)
            .build()

        gateways.get(countryCode)?.channel?.basicPublish("", createQueueName(countryCode), props, bytePayload)
    }

    private fun convertPayLoadToBytes(payload: String): ByteArray {
        val baos = ByteArrayOutputStream()
        val writter = ObjectOutputStream(baos)
        writter.writeObject(payload)
        return baos.toByteArray()
    }

    private fun createGateway(countryCode: String, host: String): Gateway {
        val gateway = Gateway(host)

        /**
         * The queue will be marked as durable to ensure that it is not lost
         * when the queue is temporarily taken down for maintenance. To delete
         * a queue created by the simulation system, please remove it by hand.
         */
        gateway.connect()
        gateway.channel?.queueDeclare(createQueueName(countryCode), true, false, false, null)

        return gateway
    }

    fun createHostProperty(countryCode: String): String = "simulation.rabbitmq.${countryCode.toLowerCase()}"

    fun createQueueName(countryCode: String): String = "rekeningrijden.simulation.${countryCode.toLowerCase()}"

    companion object {
        private val logger = LoggerFactory.getLogger(MessageProducer::class.java)
    }
}
